app/javascript/crayons/MentionAutocompleteTextArea/MentionAutocompleteTextArea.jsx

```diff
@@ -0,0 +1,263 @@
+import { h, Fragment } from 'preact';
+import { useState, useEffect, useRef, useLayoutEffect } from 'preact/hooks';
+import PropTypes from 'prop-types';
+import {
+  Combobox,
+  ComboboxInput,
+  ComboboxPopover,
+  ComboboxList,
+  ComboboxOption,
+} from '@reach/combobox';
+import '@reach/combobox/styles.css';
+import { getMentionWordData, getCursorXY } from '@utilities/textAreaUtils';
+import { useMediaQuery, BREAKPOINTS } from '@components/useMediaQuery';
+
+const MIN_SEARCH_CHARACTERS = 2;
+const MAX_RESULTS_DISPLAYED = 6;
+
+const UserListItemContent = ({ user }) => {
+  return (
+    <Fragment>
+      <span className="crayons-avatar crayons-avatar--l mr-2 shrink-0">
+        <img
+          src={user.profile_image_90}
+          alt=""
+          className="crayons-avatar__image "
+        />
+      </span>
+
+      <div>
+        <p className="crayons-autocomplete__name">{user.name}</p>
+        <p className="crayons-autocomplete__username">{`@${user.username}`}</p>
+      </div>
+    </Fragment>
+  );
+};
+
+/**
+ * A component for dynamically searching for users and displaying results in a dropdown.
+ * This component will replace the textarea passed in props, copying all styles and attributes, and allowing for progressive enhancement
+ *
+ * @param {object} props
+ * @param {element} props.replaceElement The textarea DOM element that should be replaced
+ * @param {function} props.fetchSuggestions The async call to use for the search
+ *
+ * @example
+ * <MentionAutocompleteCombobox
+ *    replaceElement={textAreaRef.current}
+ *    fetchSuggestions={fetchUsersByUsername}
+ * />
+ */
+export const MentionAutocompleteTextArea = ({
+  replaceElement,
+  fetchSuggestions,
+}) => {
+  const [textContent, setTextContent] = useState('');
+  const [searchTerm, setSearchTerm] = useState('');
+  const [cachedSearches, setCachedSearches] = useState({});
+  const [dropdownPositionPoints, setDropdownPositionPoints] = useState({
+    x: 0,
+    y: 0,
+  });
+  const [selectionInsertIndex, setSelectionInsertIndex] = useState(0);
+  const [users, setUsers] = useState([]);
+  const [cursorPosition, setCursorPosition] = useState(null);
+  const [ariaHelperText, setAriaHelperText] = useState('');
+
+  const isSmallScreen = useMediaQuery(`(max-width: ${BREAKPOINTS.Small}px)`);
+
+  const inputRef = useRef(null);
+
+  useEffect(() => {
+    if (searchTerm.length >= MIN_SEARCH_CHARACTERS) {
+      if (cachedSearches[searchTerm]) {
+        setUsers(cachedSearches[searchTerm]);
+        return;
+      }
+
+      fetchSuggestions(searchTerm).then(({ result: fetchedUsers }) => {
+        const resultLength = Math.min(
+          fetchedUsers.length,
+          MAX_RESULTS_DISPLAYED,
+        );
+
+        const results = fetchedUsers.slice(0, resultLength);
```

Comment:
We might want to make this smarter in a future PR, once #12975 is merged. We shouldn't ask the backend for all results and then slice them out in the frontend. We should simply ask `MAX_RESULTS_DISPLAYED` number of results right away using a limit or pagination. 

I'll make a note in #12975 to add that feature. Is there anything else you need from the backend results @aitchiss ?

cc @atsmith813

Reply:
Totally agree - it would be great to be able to set the max number returned in the search call. 

I don't think we need anything else from the backend results - we don't page these in or filter them at all at the moment.

Reply:
@aitchiss for now we've hardcoded the number of results in [this PR](https://github.com/forem/forem/pull/12975/files#diff-dab7116df779d917d8f1420244e812bfd0f7da4fdaaac01fb1576b925e6dbe91R4), we'll integrate a parameter if we need to and after we test the new endpoint in production.